allennlp/predictors/multitask.py

```diff
@@ -0,0 +1,102 @@
+from overrides import overrides
+from typing import Type, List, Dict
+
+from allennlp.common import JsonDict
+from allennlp.data import DatasetReader, Instance
+from allennlp.models import Model
+from allennlp.predictors.predictor import Predictor
+
+
+@Predictor.register("multitask")
+class MultiTaskPredictor(Predictor):
+    """
+    Predictor for multitask models.
+
+    Registered as a `Predictor` with name "multitask".
+    """
+
+    _WRONG_READER_ERROR = (
+        "MultitaskPredictor is designed to work with MultiTaskDatasetReader. "
+        + "If you have a different DatasetReader, you have to write your own "
+        + "Predictor, but you can use MultitaskPredictor as a starting point."
+    )
+
+    _WRONG_FIELD_ERROR = (
+        "MultiTaskPredictor expects instances that have a MetadataField "
+        + "with the name 'task', containing the name of the task the instance is for."
+    )
+
+    def __init__(self, model: Model, dataset_reader: DatasetReader) -> None:
+        from allennlp.common.checks import ConfigurationError
+        from allennlp.models.multitask import MultiTaskModel
+        from allennlp.data.dataset_readers.multitask import MultiTaskDatasetReader
+
+        if not isinstance(dataset_reader, MultiTaskDatasetReader):
+            raise ConfigurationError(self._WRONG_READER_ERROR)
+
+        if not isinstance(model, MultiTaskModel):
+            raise ConfigurationError(
+                "MultiTaskPredictor is designed to work only with MultiTaskModel."
+            )
+
+        super().__init__(model, dataset_reader)
+
+        self.predictors = {}
+        for name, head in model._heads.items():
+            predictor_name = head.default_predictor
+            predictor_class: Type[Predictor] = (
+                Predictor.by_name(predictor_name) if predictor_name is not None else Predictor  # type: ignore
+            )
+            self.predictors[name] = predictor_class(model, dataset_reader.readers[name])
+
+    @overrides
+    def predict_instance(self, instance: Instance) -> JsonDict:
+        from allennlp.data.fields import MetadataField
+        from allennlp.common.util import sanitize
+        from allennlp.data.dataset_readers import MultiTaskDatasetReader
+        from allennlp.common.checks import ConfigurationError
```

Comment:
What's the rational for putting these imports here? I think it adds unnecessary runtime overhead.

Reply:
We have problems with circular imports in many places. I work around them by using local imports for anything allennlp when I can.

This also helps with startup times.

Reply:
I *really* don't think this is good practice, especially in a function that is meant to be called repeatedly. And none of these imports should cause circular import errors in this module anyway.